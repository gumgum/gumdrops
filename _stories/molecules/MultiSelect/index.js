import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { text, select } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';

import readme from './README.md';
import MultiSelect from '../../../components/molecules/MultiSelect';

const optionsA = [
    {
        name: 'Lassie',
        value: 1,
        selected: true
    },
    {
        name: 'Snoopy',
        value: 2,
        selected: true
    },
    {
        name: 'Toto',
        value: 3,
        selected: false
    },
    {
        name: 'Brian Griffin',
        value: 3,
        selected: false
    }
];

const optionsB = [
    {
        name: 'All Pets',
        value: 0,
        selected: false
    },
    {
        name: 'Dogs',
        value: 1,
        selected: false,
        options: [
            {
                name: 'Lassie',
                value: 1,
                selected: true
            },
            {
                name: 'Snoopy',
                value: 2,
                selected: true
            },
            {
                name: 'Toto',
                value: 3,
                selected: false
            },
            {
                name: 'Brian Griffin',
                value: 3,
                selected: false
            }
        ]
    },
    {
        name: 'Cats',
        value: 'cats',
        selected: false,
        options: [
            {
                name: 'Grumpy Cat',
                value: 1,
                selected: true
            },
            {
                name: 'Lil Bub',
                value: 2,
                selected: true
            },
            {
                name: 'Hello Kitty',
                value: 3,
                selected: false
            }
        ]
    }
];

const configSelect = {
    Normal: 'A',
    Nested: 'B'
};

const options = {
    A: optionsA,
    B: optionsB
};

const sizeOptions = {
    sm: 'sm',
    xs: 'xs',
    'No Value': ''
};

class MultiSelectOptions extends Component {
    static propTypes = {
        options: PropTypes.array
    };

    state = {
        options: this.props.options
    };

    static getDerivedStateFromProps(props) {
        return {
            options: props.options
        };
    }

    handleChange = newOptions => {
        this.setState({
            options: newOptions
        });
    };

    render() {
        return (
            <MultiSelect
                placeholder={text('Placeholder', 'Select an option')}
                size={optionalSelect('Size', sizeOptions, '')}
                onChange={this.handleChange}
                options={this.state.options}
                className={text('className', '')}
            />
        );
    }
}
/* eslint-disable */
const component = () => {
    const selectedOptions = select('Options Config', configSelect, 'A');
    return <MultiSelectOptions options={options[selectedOptions]} />;
};
/* eslint-enable */

export default [readme, component];
