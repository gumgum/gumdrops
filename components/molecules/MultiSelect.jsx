import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import MultiSelectSubMenu from './MultiSelectSubMenu';
import MultiSelectMenuItem from './MultiSelectMenuItem';

import generateUID from '../utils/generateUID';
import updateOptions from '../utils/updateMultiSelectOptions';
import cx from 'classnames';


const getSelectedOptions = (options) => {
  const selected = options.reduce((acc, cur) => {
    if (cur.options) {
      return [...acc, ...getSelectedOptions(cur.options)];
    }
    return cur.selected ? [...acc, cur] : acc;
  }, []);

  return selected;
};

const MultiSelect = ({
  options,
  onChange,
  size,
  className,
  ulStyle,
  ...otherProps
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const containerRef = useRef();

  const uid = generateUID(containerRef);

  const regionId = `MultiSelect_region_${uid}`;
  const labelId = `MultiSelect_label_${uid}`;

  useEffect(() => {
    const closeOnClickOutside = ({ target }) => {
      const el = containerRef.current;
      if (!el.contains(target)) {
        setIsOpen(false);
      }
    };

    window.addEventListener('click', closeOnClickOutside);

    return () => {
      window.removeEventListener('click', closeOnClickOutside);
    };
  }, []);

  const toggleDropdown = () => setIsOpen((prevState) => !prevState);

  const getPlaceholder = () => {
    const { placeholder } = otherProps;
    const selectedOptions = getSelectedOptions(options);
    return selectedOptions.map((x) => x.name).join(', ') || placeholder;
  };

  const handleSubChange = (index) => (subItem) => {
    const updatedOptions = options.map((item, i) => {
      const shouldUpdate = i === index;
      if (shouldUpdate) {
        const subOptions = updateOptions(item.options, subItem);
        if (subOptions) {
          const allSelected = !subOptions.map((opt) => opt.selected).includes(false);
          return {
            ...item,
            selected: allSelected,
            options: subOptions
          };
        }
      }
      return item;
    });

    onChange(updatedOptions);
  };

  const handleChange = (item) => {
    const updatedOptions = updateOptions(options, item);
    onChange(updatedOptions);
  };

  const { options: _, ...restProps } = otherProps;

  const isSmall = size === 'sm';
  const isExtraSmall = size === 'xs';

  const rootClass = cx('gds-multi-select', className, {
    'gds-multi-select--sm': isSmall,
    'gds-multi-select--xs': isExtraSmall,
    'gds-button-dropdown--active': isOpen
  });

  const btnClass = cx('gds-multi-select__button', {
    'gds-multi-select__button--sm': isSmall,
    'gds-multi-select__button--xs': isExtraSmall
  });

  return (
      <div className={rootClass} {...restProps} ref={containerRef}>
          <button
              aria-expanded={isOpen}
              aria-pressed={isOpen}
              aria-controls={regionId}
              tabIndex={0}
              className={btnClass}
              id={labelId}
              name="multiselectMenu"
              type="button"
              onClick={toggleDropdown}>
              <div className="-ellipsis">{getPlaceholder()}</div>
          </button>
          <ul
              aria-labelledby={labelId}
              aria-hidden={!isOpen}
              className="gds-multi-select__menu"
              id={regionId}
              role="region"
              style={isOpen ? ulStyle : {}}>
              {options.map(({ name, value, selected, options: subOptions }, index) =>
                  subOptions ? (
                      <MultiSelectSubMenu
                          key={`menu-item-${index}`}
                          name={name}
                          index={index}
                          value={value}
                          selected={selected}
                          options={subOptions}
                          onChange={handleChange}
                          onSubChange={handleSubChange(index)}
                          size={size}
                      />
                  ) : (
                      <MultiSelectMenuItem
                          key={`menu-item-${index}`}
                          name={name}
                          index={index}
                          value={value}
                          selected={selected}
                          onChange={handleChange}
                          size={size}
                      />
                  )
              )}
          </ul>
      </div>
  );
};

MultiSelect.displayName = 'MultiSelect';

MultiSelect.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      value: PropTypes.any,
      selected: PropTypes.bool,
      options: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string.isRequired,
          value: PropTypes.any,
          selected: PropTypes.bool
        })
      )
    })
  ).isRequired,
  callback: PropTypes.func,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  size: PropTypes.oneOf(['xs', 'sm', '']),
  className: PropTypes.string,
  ulStyle: PropTypes.object,
};

MultiSelect.defaultProps = {
  placeholder: '',
  className: '',
  onChange: () => {}, // Placeholder for the default onChange prop
  ulStyle: {}
};

export default MultiSelect;
