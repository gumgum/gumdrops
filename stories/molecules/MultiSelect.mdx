import { Story, Source, Props } from '@storybook/addon-docs/blocks';
import MultiSelect from '../../components/molecules/MultiSelect';

# MultiSelect

The `<MultiSelect>` component can be used when you need a select component to support multiple options simultaneously.

The state of the multi-select is handled by the parent component. An `onChange` prop handler is provided that will be called with the updated list of options whenever there's a change.

Multi-select also supports nested options [See Nested Options](#nested-options). When a parent option is selected, all children will be selected. If all children are selected, the parent will be selected too.

**NOTE**: This is a controlled component, so don't forget to pass the updated props to `MultiSelect` to see the changes. [See Example](#example)

## Options

Options determine the construction of the multi-select menu. It takes an `Array` of `Objects` with the following shape:

```js
{
    name: 'John', // label displayed next to the checkbox
    value: 'john', // value associated with the option
    selected: true // current selected state
}
```

Here's an example of a possible set of options:

```js
const people = [
    {
        name: 'John',
        value: 'john',
        selected: true
    },
    {
        name: 'Sam',
        value: 'sam',
        selected: false
    },
    {
        name: 'Jane',
        value: 'jane',
        selected: true
    },
    {
        name: 'Sara',
        value: 'sara',
        selected: false
    }
];
```

### Nested Options

You may also have nested options. Here's an example:

```js
const pets = [
    {
        name: 'All Pets',
        value: 0,
        selected: false
    },
    {
        name: 'Dogs',
        value: 1,
        selected: false,
        options: [
            {
                name: 'Lassie',
                value: 1,
                selected: true
            },
            {
                name: 'Snoopy',
                value: 2,
                selected: true
            }
        ]
    },
    {
        name: 'Cats',
        value: 'cats',
        selected: false,
        options: [
            {
                name: 'Grumpy Cat',
                value: 1,
                selected: true
            },
            {
                name: 'Lil Bub',
                value: 2,
                selected: true
            }
        ]
    }
];
```

## Full Example

```jsx
const [pets, setPets] = useState(initialPetsObject);
return <MultiSelect onChange={newPets => setPets(newPets)} options={pets} size="md" />;
```

## Component Example

<Story id="molecules-multiselect--default" />

## Source

<Source id="molecules-multiselect--default" />

## Props

<Props of={MultiSelect} />
