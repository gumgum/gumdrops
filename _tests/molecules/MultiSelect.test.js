/* globals mount, shallow */
import React from 'react';
import MultiSelect from '../../components/molecules/MultiSelect';

const options = [
    { name: 'All Pets', value: 0, selected: false },
    {
        name: 'Dogs',
        value: 1,
        selected: false,
        options: [
            { name: 'Lassie', value: 1, selected: true },
            { name: 'Snoopy', value: 2, selected: true },
            { name: 'Toto', value: 3, selected: false },
            { name: 'Brian Griffin', value: 3, selected: false }
        ]
    },
    {
        name: 'Cats',
        value: 'cats',
        selected: false,
        options: [
            { name: 'Grumpy Cat', value: 1, selected: true },
            { name: 'Lil Bub', value: 2, selected: true },
            { name: 'Hello Kitty', value: 3, selected: false }
        ]
    }
];

const defaultProps = {
    options,
    callback: () => {},
    placeholder: 'placeholder',
    size: 'xs',
    className: 'custom-class'
};

describe('Expect <MultiSelect>', () => {
    it('to render', () => {
        const wrapper = mount(<MultiSelect {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('to open and close', () => {
        const wrapper = shallow(<MultiSelect {...defaultProps} />);
        const menuBtn = wrapper.find('button[name="multiselectMenu"]');
        expect(wrapper.state().isOpen).toEqual(false);
        // open
        menuBtn.simulate('click');
        expect(wrapper.state().isOpen).toEqual(true);
        expect(wrapper).toMatchSnapshot();
        // close
        menuBtn.simulate('click');
        expect(wrapper.state().isOpen).toEqual(false);
    });
});
