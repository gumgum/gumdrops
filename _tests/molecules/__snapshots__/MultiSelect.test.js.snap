// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <MultiSelect> to render properly 1`] = `
<div
  callback={[MockFunction]}
  className="gds-multi-select custom-class gds-multi-select--xs"
  placeholder="placeholder"
>
  <button
    aria-controls="MultiSelect_region_T2JqZWN00"
    aria-expanded={false}
    aria-pressed={false}
    className="gds-multi-select__button gds-multi-select__button--xs"
    id="MultiSelect_label_T2JqZWN00"
    name="multiselectMenu"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="-ellipsis"
    >
      Lassie, Snoopy, Grumpy Cat, Lil Bub
    </div>
  </button>
  <ul
    aria-hidden={true}
    aria-labelledby="MultiSelect_label_T2JqZWN00"
    className="gds-multi-select__menu"
    id="MultiSelect_region_T2JqZWN00"
    role="region"
    style={{}}
  >
    <li
      className="gds-multi-select__menu-item"
    >
      <div
        className="gds-multi-select__menu-link"
      >
        <div
          className="gds-form-group gds-multi-select__option"
          onClick={[Function]}
        >
          <div
            className="gds-form-group__checkbox gds-form-group__checkbox--xs"
          >
            <label
              className="gds-form-group__checkbox-label"
            >
              <input
                checked={false}
                className="gds-form-group__checkbox-input"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="gds-form-group__checkbox-indicator"
              />
              All Pets
            </label>
          </div>
        </div>
      </div>
    </li>
    <li
      className="gds-multi-select__menu-item"
      onClick={[Function]}
    >
      <div
        className="gds-multi-select__menu-link gds-multi-select__menu-link--collapsable gds-multi-select__menu-link--active"
      >
        <div
          className="gds-form-group gds-multi-select__option"
          onClick={[Function]}
        >
          <div
            className="gds-form-group__checkbox gds-form-group__checkbox--xs"
          >
            <label
              className="gds-form-group__checkbox-label"
            >
              <input
                checked={false}
                className="gds-form-group__checkbox-input"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="gds-form-group__checkbox-indicator"
              />
              Dogs
            </label>
          </div>
        </div>
        <ul
          className="gds-multi-select__sub-menu"
        >
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={true}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Lassie
                  </label>
                </div>
              </div>
            </div>
          </li>
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={true}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Snoopy
                  </label>
                </div>
              </div>
            </div>
          </li>
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={false}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Toto
                  </label>
                </div>
              </div>
            </div>
          </li>
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={false}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Brian Griffin
                  </label>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </li>
    <li
      className="gds-multi-select__menu-item"
      onClick={[Function]}
    >
      <div
        className="gds-multi-select__menu-link gds-multi-select__menu-link--collapsable gds-multi-select__menu-link--active"
      >
        <div
          className="gds-form-group gds-multi-select__option"
          onClick={[Function]}
        >
          <div
            className="gds-form-group__checkbox gds-form-group__checkbox--xs"
          >
            <label
              className="gds-form-group__checkbox-label"
            >
              <input
                checked={false}
                className="gds-form-group__checkbox-input"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="gds-form-group__checkbox-indicator"
              />
              Cats
            </label>
          </div>
        </div>
        <ul
          className="gds-multi-select__sub-menu"
        >
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={true}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Grumpy Cat
                  </label>
                </div>
              </div>
            </div>
          </li>
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={true}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Lil Bub
                  </label>
                </div>
              </div>
            </div>
          </li>
          <li
            className="gds-multi-select__menu-item"
            onClick={[Function]}
          >
            <div
              className="gds-multi-select__menu-link"
            >
              <div
                className="gds-form-group gds-multi-select__option"
              >
                <div
                  className="gds-form-group__checkbox gds-form-group__checkbox--xs"
                >
                  <label
                    className="gds-form-group__checkbox-label"
                  >
                    <input
                      checked={false}
                      className="gds-form-group__checkbox-input"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <span
                      className="gds-form-group__checkbox-indicator"
                    />
                    Hello Kitty
                  </label>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>
`;
